                                                               
                                                               
                                                               
                                                                #############################
                                                                ### Metagenomic profiling ###
                                                                #############################


########################################################################
### Quality Check and filtering of human reads with KneadData 0.7.10 ###
########################################################################

kneaddata \
--input ./sampleX_read1.fq.gz \
--input ./sampleX_read2.fq.gz \
--reference-db ./hg37dec_v0.1 \
--output ./sampleX \
--threads 12 \
--processes 12 \
--run-fastqc-start \
--run-fastqc-end \
--bypass-trf \
--sequencer-source TruSeq3

################################################
### Taxonomic Profiling with MetaPhlAn 3.0.2 ###
################################################

# As input, the paired output files generated by KneadData after QC and filtering of human reads were used
metaphlan \
./sampleX_paired_1.fastq,./sampleX_paired_2.fastq \
--nproc 12 \
--input_type fastq \
-o ./sampleX_profiled.tsv

# merge abundance tables
merge_metaphlan_tables.py path/to/*_profiled.tsv > ./rarefied_profiled_merged.tsv

# subset to species level (bash)
grep -E "s__|clade" ./rarefied_profiled_merged.tsv | sed 's/^.*s__//g'\ | cut -f1,3-104 | sed -e 's/clade_name/sample/g' > ./rarefied_profiled_merged_species.tsv

######################################################
#### Rarefaction metagenomic reads to common depth ###
######################################################

# For diversity analysis, metagenomic samples were rarefied to increasing read depths prior to taxonomic assignemt with MetaPhlAn
# The variable $readDepth represents increasing read depths
# As input, the paired output files generated by KneadData after QC and filtering of human reads were used
seqkit sample \
./sampleX_paired_1.fastq \
-n $readDepth \
-s 11 \
-j 12 \
-t dna \
-o ./sampleX_rarefied_read1.fastq

seqkit sample \
./sampleX_paired_1.fastq \
-n $readDepth \
-s 11 \
-j 12 \
-t dna \
-o ./sampleX_rarefied_read2.fastq

###############################################
### Virulence gene profiling with ShortBRED ###
###############################################

# Use ShortBRED-Identify to identify markers for proteins of interest
python shortbred_identify.py \
--goi  ./VFDB_setA_proCoreDB.faa \
--ref ./uniref90.fasta \
--markers ./VFDB_setA_proCoreDB_markers.faa \
--threads 12

# Use ShortBRED-Quantify to quantify the abundance of proteins of interest in the metagenomic sample
# As input, the paired output files generated by KneadData after QC and filtering of human reads were used
python shortbred_quantify.py \
--markers VFDB_setA_proCoreDB_markers.faa \
--wgs sampleX_paired_1.fastq sampleX_paired_2.fastq sampleX_unmatched_1.fastq sampleX_unmatched_2.fastq \
--results sampleX_results.txt

##########################################
### Strain Profiling with PanPhlAn 3.1 ###
##########################################

# Download pangenomes of interest
panphlan_download_pangenome.py \
-i Staphylococcus_aureus
panphlan_download_pangenome.py \
-i Staphylococcus_hominis
panphlan_download_pangenome.py \
-i Staphylococcus_epidermidis
panphlan_download_pangenome.py \
-i Cutibacterium_acnes

# Map each metagenomic sample against each species' pangenenome (shown is an example for sampleX and Staphylococcus aureus)
# As input, the paired output files generated by KneadData after QC and filtering of human reads were used.
# Read 1 and read 2 were concatenated
read1 and read2 were concatenated beforehand
panphlan_map.py \
-i sampleX_paired_cat.fastq \
--indexes ./Staphylococcus_aureus/Staphylococcus_aureus \
-p ./Staphylococcus_aureus/Staphylococcus_aureus_pangenome.tsv \
-o ./sampleX_aureus.tsv \
--verbose \
-m 30

# Process and merge the mapping results from each metagenomic sample
python panphlan/panphlan_profiling.py \
-i path/to/dir/with/results/for/Staphylococcus_aureus \
--o_matrix result_profile_Saureus_least_stringent.tsv \
-p ./Staphylococcus_aureus/Staphylococcus_aureus_pangenome.tsv \
--min_coverage 1 \
--left_max 1.70 \
--right_min  0.30

#############################################################################
### Rarefaction of Metagenomic Samples and subsequent Taxonomic Profiling ###
#############################################################################

# Subsample metagenomic reads to common depth (python)
seqkit sample \
./sampleX_read1.fastq \
-n $readDepth \
-s 11 \
-j 12 \
-t dna \
-o ./sampleX_rarefied_read1.fastq

# Profile with MetaPhlAn (python)
metaphlan sampleX_rarefied_read1.fastq,sampleX__rarefied_read1.fastq --nproc 12 --input_type fastq -o sampleX_rarefied_profiled.tsv

# Merge abundance tables (python)
merge_metaphlan_tables.py path/to/*_profiled.tsv > rarefied_profiled_merged.tsv

Subset to species level(bash)
grep -E "s__|clade" rarefied_profiled_merged.tsv | sed 's/^.*s__//g'\ | cut -f1,3-104 | sed -e 's/clade_name/sample/g' > rarefied_profiled_merged_species.tsv